### ISSUE TITLE
Restrict which keys can be accessed when reading

### ISSUE THREAD

**Issue by ReeceBlack123** [hashicorp/vault#15951]
**Is your feature request related to a problem? Please describe.**

I would like to read specific keys that start with certain prefix e.g `AKIA`

**Describe the solution you'd like**

There is `allowed_parameters` policy option to restrict which keys are allowed to be created.
Is there something similar that is available to restrict which keys can be read depending on the policy available to the token?
If not could `allowed_parameters` be extended to `read`?

**Comment by raskchanky** [comment]
Hi @ReeceBlack123 

I don't believe that anything exists right now that will do exactly what you want. If you're using kv-v2, there is a capability that is similar, but not quite the same thing: https://www.vaultproject.io/api-docs/secret/kv/kv-v2#read-secret-subkeys

**Comment by maxb** [comment]
> I would like to read specific keys that start with certain prefix e.g `AKIA`

Please explain further - I thought I understood what you wanted - selective reads of subkeys within a KV value... but `AKIA` is the prefix of AWS access key IDs, which is something else entirely, and would not typically be present as a key, but a value.



> If not could `allowed_parameters` be extended to `read`?

`allowed_parameters` already has meaning for `read` - it restricts which parameters can be supplied in the request.

Currently the ACL policy subsystem of Vault is strictly an allow-or-deny filter. If new functionality was to be added to Vault that allowed filtering of the **content of responses** by the ACL policy subsystem, then it would be terribly confusing to override the existing `allowed_parameters` construct and it would need to be configured by some other setting.

**Comment by ReeceBlack123** [comment]
> Please explain further - I thought I understood what you wanted - selective reads of subkeys within a KV value... but `AKIA` is the prefix of AWS access key IDs, which is something else entirely, and would not typically be present as a key, but a value.

At the moment, there are multiple applications that are using AWS access key IDs.  I am trying to find all AWS access key IDs in the path structure: `/secrets/{application-name/{environment}`

Example:
- /secrets/app-1/production`
- /secrets/app-2/test`
- /secrets/app-3/test`

I am trying to create a policy for an [app role](https://www.vaultproject.io/docs/auth/approle) that would only have permissions to read AWS access key IDs in the path described above because I don't want the app role to have access to other secrets. 

Is there a way for the app role to have access to all AWS access key IDs in the path structure described above?

------------------------------------------------------------------------------
Additional thought:
Some applications aren't using the usual AWS access key id key (`AWS_ACCESS_KEY_ID`). 
That is why I was thinking of only allowing the app role to only have access to read values with `AKIA` prefix. Do you know if that is possible?

**Comment by maxb** [comment]
Thanks, it is now clearer what you're looking to achieve - not only are you interested in filtering what is returned by the **keys** in the JSON payload (e.g. to keys equal to `AWS_ACCESS_KEY_ID`), but you are also interested in selectively revealing certain **values** within the JSON payload too, based on whether or not they match a pattern (`AKIA*`).

I would characterise these as two different feature requests.

Neither of these capabilities exist in Vault today.

For your immediate needs, I think you will need to build and deploy a privileged application which can use highly sensitive credentials with read access to `/secrets/*`, and return redacted information to its users.

**Comment by ReeceBlack123** [comment]
> For your immediate needs, I think you will need to build and deploy a privileged application which can use highly sensitive credentials with read access to /secrets/*, and return redacted information to its users.

Thanks. How does returning redacted information to its users happen? Is this a built-in feature in Vault are there any docs?

**Comment by maxb** [comment]
No, I mean, this is an application you would have to write yourself.
